### ISSUE TITLE
`no-array-for-each`: False positive with Maps/Sets

### ISSUE THREAD

**Issue by delucis** [sindresorhus/eslint-plugin-unicorn#1452]
When the `no-array-for-each` rule is enabled (as recommended) it reports errors when `forEach` is used with other types that aren’t arrays, like `Map` or `Set`. (Or any custom object with a `forEach` method potentially.)

```js
const map = new Map();
map.forEach((i) => console.log(i));
//  ^^^^^^^
//  Error: Do not use `Array#forEach(…)`

const set = new Set();
set.forEach((i) => console.log(i));
//  ^^^^^^^
//  Error: Do not use `Array#forEach(…)`
```

This is most important for `Map`, where `forEach` passes the mapped value as the first argument to its callback, but the current auto-fix (`for...of`) will pass `[key, value]`:

```js
const map = new Map([['key', 'value']]);

// Before auto-fix
map.forEach((value) => console.log(value)); // 'value'

// After auto-fix
for (const value of map) {
  console.log(value); // ['key', 'value']
}
```

**Comment by fisker** [comment]
The  `Set` case should be safe, but we don't know if the object is a `Map`, I guess we'll have to turn auto-fix into a suggestion. Or maybe assume it's not `Map` under an option.

**Comment by clshortfuse** [comment]
Considering there's no way to disable autofix on eslint, it makes the rule dangerous to use at all, even as a warning.

**Comment by AlexAegis** [comment]
I just run into this, auto-fixing broke a lot of Map.forEach-s.

This rule should be just removed as it cannot do it's job properly without type information.

**Comment by axetroy** [comment]
The rule should probably only autofix when the argument type is `Array`. Otherwise, there are no errors and suggestions should be prompted

**Comment by fregante** [comment]
> The `Set` case should be safe, but we don't know if the object is a `Map`, I guess we'll have to turn auto-fix into a suggestion. Or maybe assume it's not `Map` under an option.

I'd say the rule should:

- [ ] auto-fix only when the type is known
- [ ] add auto-fix support for `Map` when detected

The rule itself is still applicable in all cases described here, the only issue is the autofix
